internal/rbd/encryption.go

```diff
@@ -175,21 +184,63 @@ func (rv *rbdVolume) initKMS(ctx context.Context, volOptions, credentials map[st
 		return nil
 	}
 
-	rv.Encrypted, err = strconv.ParseBool(encrypted)
+	isEncrypted, err := strconv.ParseBool(encrypted)
 	if err != nil {
 		return fmt.Errorf(
 			"invalid value set in 'encrypted': %s (should be \"true\" or \"false\")", encrypted)
-	} else if !rv.Encrypted {
+	} else if !isEncrypted {
 		return nil
 	}
 
-	// deliberately ignore if parsing failed as GetKMS will return default
-	// implementation of kmsID is empty
-	kmsID := volOptions["encryptionKMSID"]
-	rv.KMS, err = util.GetKMS(rv.Owner, kmsID, credentials)
+	err = rv.configureEncryption(volOptions["encryptionKMSID"], credentials)
 	if err != nil {
 		return fmt.Errorf("invalid encryption kms configuration: %w", err)
 	}
 
 	return nil
 }
+
+func (rv *rbdVolume) configureEncryption(kmsID string, credentials map[string]string) error {
+	kms, err := util.GetKMS(rv.Owner, kmsID, credentials)
+	if err != nil {
+		return err
+	}
+
+	rv.encryption, err = util.NewVolumeEncryption(kms)
+
+	// if the KMS can not store the DEK itself, we'll store it in the
+	// metadata of the RBD image itself
+	if errors.Is(err, util.ErrDEKStoreNeeded) {
+		rv.encryption.SetDEKStore(rv)
+	}
+
+	return nil
+}
+
+// StoreDEK saves the DEK in the metadata, overwrites any existing contents.
```

Comment:
nit: may be we can mention rbd image metadata here to make it more clear.

Reply:
It's the method of the `rbdVolume` type, so it should be clear this is the metadata of the rbd-volume?

Reply:
I am fine either way..